### PR TITLE
Scan fields from ES constructor declarations

### LINKED ISSUE #706: [Feature Request] Collect instance properties set in class constructor
I'd like to be able to collect instance properties for classes to help with closure interface generation with https://github.com/Polymer/gen-closure-declarations.
To do that in an expedient way, it would be enough to just collect assignments or references to instance properties from the `constructor`.

### PR DESCRIPTION
**Pull request by rictic**
- [x] CHANGELOG.md has been updated

This is a long awaited feature. @azakus is the latest to be pushing for it.

Fixes #706

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -19,7 +19,7 @@
   ],
   "scripts": {
     "clean": "mkdir -p lib/ && rm -rf lib/",
-    "build": "gulp build",
+    "build": "gulp build --silent",
```
Comment (justinfagnani) [inline package.json]: why? can we get rid of gulp yet?

**Comment by azakus** [comment]
Seems to work well for me in gen-closure-declarations!

File: src/javascript/class-scanner.ts [hunk 2]
```diff
@@ -574,3 +585,93 @@ class CustomElementsDefineCallFinder implements Visitor {
     return null;
   }
 }
+
+export function extractPropertiesFromConstructor(
+    astNode: estree.Node,
+    document: JavaScriptDocument,
+    properties = new Map<string, ScannedProperty>()) {
+  if (astNode.type === 'ClassExpression' ||
```
Comment (justinfagnani) [inline src/javascript/class-scanner.ts]: seems a good bail-early place

File: src/javascript/class-scanner.ts [hunk 3]
```diff
@@ -574,3 +585,93 @@ class CustomElementsDefineCallFinder implements Visitor {
...
+          continue;
+        }
+        let name;
+        let astNode;
+        let defaultValue;
+        if (statement.expression.type === 'AssignmentExpression') {
+          name = getPropertyNameOnThisExpression(statement.expression.left);
+          astNode = statement.expression.left;
+          defaultValue = escodegen.generate(statement.expression.right);
+        } else if (statement.expression.type === 'MemberExpression') {
```
Comment (justinfagnani) [inline src/javascript/class-scanner.ts]: can you document his branch? I assume you looking for just `/** @public */ this.foo;`

File: src/polymer/polymer-element.ts [hunk 4]
```diff
@@ -43,8 +44,78 @@ export interface BasePolymerProperty {
...
+    propB: Readonly<ScannedPolymerProperty>): ScannedPolymerProperty {
+  if (propA.name !== propB.name) {
+    throw new Error(
+        `Tried to merge properties with different names: ` +
+        `'${propA.name}' and ' ${propB.name}'`);
+  }
+  const name = propA.name;
+  let description;
+  if (propA.description !== undefined && propB.description !== undefined) {
+    description = propA.description.length < propB.description.length ?
```
Comment (justinfagnani) [inline src/polymer/polymer-element.ts]: I hate this logic, but don't have a better idea. Can we put it in a common location so we can share it and document it once?

File: package.json (hunk 1 above)
Comment (rictic) [inline package.json]: it copies the static test fixtures over, as well as generating the json schema from typescript types.

File: src/javascript/class-scanner.ts (hunk 2 above)
Comment (rictic) [inline src/javascript/class-scanner.ts]: +1, done

File: src/javascript/class-scanner.ts (hunk 3 above)
Comment (rictic) [inline src/javascript/class-scanner.ts]: Yep, that's right. Done.

File: src/polymer/polymer-element.ts (hunk 4 above)
Comment (rictic) [inline src/polymer/polymer-element.ts]: Moved to jsdoc. It duplicates some of `jsdoc.join(...)`. I remember that there was a pretty good argument for that logic.. I think. But this looks correct for this case. Left a comment to investigate.

**Pull request merged**